File: app/src/test/java/com/nicholasworkshop/tinklabstest/widget/AdItemViewTest.java
package com.nicholasworkshop.tinklabstest.widget;

import android.content.Context;
import android.content.Intent;
import android.widget.ImageView;

import com.bumptech.glide.DrawableTypeRequest;
import com.bumptech.glide.RequestManager;
import com.github.javafaker.Faker;
import com.nicholasworkshop.tinklabstest.MainApplication;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.Spy;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;
import org.robolectric.annotation.Config;

import static org.junit.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static org.mockito.MockitoAnnotations.initMocks;

/**
 * Created by nicholas.wong on 2017/05/16.
 */
@RunWith(RobolectricTestRunner.class)
@Config(manifest = Config.NONE, application = MainApplication.class)
public class AdItemViewTest {

    private final Faker faker = new Faker();

    @Spy Context context = RuntimeEnvironment.application;
    @Mock ImageView adImageView;
    @Mock RequestManager requestManager;
    @Mock DrawableTypeRequest<String> drawableTypeRequest;
    @Captor ArgumentCaptor<Intent> intentCaptor;

    private AdItemView adItemView;

    @Before
    public void setUp() throws Exception {
        initMocks(this);
        when(requestManager.load(anyString())).thenReturn(drawableTypeRequest);
        adItemView = new AdItemView(context);
        adItemView.requestManager = requestManager;
        adItemView.adImageView = adImageView;
    }

    @Test
    public void setImageUrl() throws Exception {
        String url = faker.internet().image();
        adItemView.setImageUrl(url);
        //////
        verify(requestManager).load(url);
        verify(drawableTypeRequest).into(adImageView);
    }

    @Test
    public void clear() throws Exception {
        adItemView.setImageUrl(faker.internet().image());
        adItemView.setRedirectUrl(faker.internet().url());
        adItemView.clear();
        //////
        verify(adImageView).setImageDrawable(null);
    }

    @Test
    public void onClicked() throws Exception {
        String url = faker.internet().url();
        adItemView.setRedirectUrl(url);
        adItemView.onClicked();
        ///////
        verify(context).startActivity(intentCaptor.capture());
        Intent intent = intentCaptor.getValue();
        assertEquals(Intent.ACTION_VIEW, intent.getAction());
        assertEquals(Intent.FLAG_ACTIVITY_NEW_TASK, intent.getFlags());
        assertEquals(url, intent.getData().toString());
    }

    @Test
    public void onClicked_whenRedirectUrlNotSet_thenNothingHappens() throws Exception {
        adItemView.onClicked();
        //////
        verify(context, never()).startActivity(any());
    }

    @Test
    public void onClicked_whenCleared_thenNothingHappens() throws Exception {
        adItemView.setImageUrl(faker.internet().image());
        adItemView.setRedirectUrl(faker.internet().url());
        adItemView.clear();
        adItemView.onClicked();
        //////
        verify(context, never()).startActivity(any());
    }
}

File: app/src/test/kotlin/com/nicholasworkshop/tinklabstest/widget/AdItemViewTest.kt
package com.nicholasworkshop.tinklabstest.widget

import android.content.Context
import android.content.Intent
import android.graphics.drawable.Drawable
import android.widget.ImageView
import com.bumptech.glide.RequestBuilder
import com.bumptech.glide.RequestManager
import com.github.javafaker.Faker
import com.nhaarman.mockito_kotlin.whenever
import com.nicholasworkshop.tinklabstest.MainApplication
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentCaptor
import org.mockito.ArgumentMatchers.any
import org.mockito.ArgumentMatchers.anyString
import org.mockito.Captor
import org.mockito.Mock
import org.mockito.Mockito.never
import org.mockito.Mockito.verify
import org.mockito.MockitoAnnotations.initMocks
import org.mockito.Spy
import org.robolectric.RobolectricTestRunner
import org.robolectric.RuntimeEnvironment
import org.robolectric.annotation.Config

@RunWith(RobolectricTestRunner::class)
@Config(manifest = Config.NONE, application = MainApplication::class)
class AdItemViewTest {

    private val faker = Faker()

    @Spy private var context: Context = RuntimeEnvironment.application

    @Mock private lateinit var adView: ImageView
    @Mock private lateinit var requestManager: RequestManager
    @Mock private lateinit var requestBuilder: RequestBuilder<Drawable>

    @Captor private lateinit var intentCaptor: ArgumentCaptor<Intent>

    private lateinit var adItemView: AdItemView

    @Before
    fun setUp() {
        initMocks(this)
        whenever(requestManager.load(anyString())).thenReturn(requestBuilder)
        adItemView = AdItemView(context)
        adItemView.requestManager = requestManager
        adItemView.adView = adView
    }

    @Test
    fun setImageUrl() {
        val url = faker.internet().image()
        adItemView.setImageUrl(url)
        verify(requestManager).load(url)
        verify(requestBuilder).into(adItemView.adView)
    }

    @Test
    fun clear() {
        adItemView.setImageUrl(faker.internet().image())
        adItemView.redirectUrl = faker.internet().url()
        adItemView.clear()
        assertNull(adItemView.adView.drawable)
        verify(adView).setImageDrawable(null)
    }

    @Test
    fun onClicked() {
        val url = faker.internet().url()
        adItemView.redirectUrl = url
        adItemView.performClick()
        verify(context).startActivity(intentCaptor.capture())
        val intent = intentCaptor.value
        assertEquals(Intent.ACTION_VIEW, intent.action)
        assertEquals(Intent.FLAG_ACTIVITY_NEW_TASK.toLong(), intent.flags.toLong())
        assertEquals(url, intent.data.toString())
    }

    @Test
    fun onClicked_whenRedirectUrlNotSet_thenNothingHappens() {
        adItemView.performClick()
        verify(context, never()).startActivity(any())
    }

    @Test
    fun onClicked_whenCleared_thenNothingHappens() {
        adItemView.setImageUrl(faker.internet().image())
        adItemView.redirectUrl = faker.internet().url()
        adItemView.clear()
        adItemView.performClick()
        verify(context, never()).startActivity(any())
    }
}